### PR TITLE
First draft of BPI directory - WIP

### PR DESCRIPTION
**Pull request by erikmwilson**
The architecture is set up to have a folder that publishes externally, a folder that publishes internally, and a common folder for files that are both for internal and external consumption.

### CODE REVIEW

**Comment by erikmwilson** [comment]
@nimbinatus @robb-romans @kmsholcomb First of all, will this scheme work? Tox choked on me so I know something is wrong but I don't know what. Can you also check the plumbing for both the internal and external publishing? I'm not sure how to test those.

File: BPI-user-guide/doc/BPI-user-guide/internal/Makefile0 [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Makefile for Sphinx documentation
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/internal/Makefile0]: So this file was expected to drop into a `docs/` directory with the content (and named simply `Makefile`). I was basing that structure off of our current repo system. If we want to have a set of `internal` and `external` directories that house the content, I probably need to tweak the internal Jenkins system. Thoughts?

File: BPI-user-guide/doc/BPI-user-guide/internal/build.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/internal/build.sh]: Did you copy this file or copy-paste the text into a new file? If it's a new file with the text copy-pasted, we need to make the file executable (`sudo chmod +x <filename>.sh` on the command line) before Jenkins can use it. This note is going into the internal tools readme in the appropriate PR.

File: BPI-user-guide/doc/BPI-user-guide/internal/conf.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
+# -*- coding: utf-8 -*-
+"""
+<OFFICIAL PROJECT NAME> documentation build configuration file.
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/internal/conf.py]: This line needs to be updated.

File: BPI-user-guide/doc/BPI-user-guide/internal/publish.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/internal/publish.sh]: Did you copy this file or copy-paste the text into a new file? If it's a new file with the text copy-pasted, we need to make the file executable (`sudo chmod +x <filename>.sh` on the command line) before Jenkins can use it. This note is going into the internal tools readme in the appropriate PR.

File: BPI-user-guide/doc/BPI-user-guide/internal/test.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/internal/test.sh]: Did you copy this file or copy-paste the text into a new file? If it's a new file with the text copy-pasted, we need to make the file executable (`sudo chmod +x <filename>.sh` on the command line) before Jenkins can use it. This note is going into the internal tools readme in the appropriate PR.

File: BPI-user-guide/doc/BPI-user-guide/internal/variables.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#! /bin/sh
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/internal/variables.sh]: Did you copy this file or copy-paste the text into a new file? If it's a new file with the text copy-pasted, we need to make the file executable (`sudo chmod +x <filename>.sh` on the command line) before Jenkins can use it. This note is going into the internal tools readme in the appropriate PR.

File: BPI-user-guide/doc/BPI-user-guide/external/conf.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+# Exclude content from elastic search index
+deconst_default_unsearchable = True
+
+# linkcheck options
+linkcheck_ignore = ['https://pages.github.rackspace.com*',
+                    'https://pitchfork.eco.rackspace.com*']
+linkcheck_anchors = False
+
+# General information about the project.
+project = 'Rackspace product'
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/external/conf.py]: This line needs to be updated.

File: BPI-user-guide/doc/BPI-user-guide/external/conf.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+copyright = 'date'
```
Comment (nimbinatus) [inline BPI-user-guide/doc/BPI-user-guide/external/conf.py]: Ditto

**Comment by erikmwilson** [comment]
@nimbinatus Does the above directory allow for internal and external publishing. I am trying to set this up like the RPC publishing (https://github.com/rackerlabs/docs-rpc/tree/master/doc) which publishes both ways.

File: BPI-user-guide/doc/BPI-user-guide/internal/conf.py (hunk 3 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/conf.py]: Done

File: BPI-user-guide/doc/BPI-user-guide/internal/conf.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+# The master toctree document.
+master_doc = 'index'
+
+# The builder to use when running via the deconst preparer
+# builder = 'deconst-single'
+# builder = 'deconst-serial'
+
+# General information about the project.
+project = '<OFFICIAL PROEJCT NAME>'
```
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/conf.py]: Done.

File: BPI-user-guide/doc/BPI-user-guide/internal/conf.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+copyright = '<YEAR>, Rackspace'
```
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/conf.py]: Done.

File: BPI-user-guide/doc/BPI-user-guide/internal/Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+def isPR=false
+
+// Staging post
+def notifyGitHub(text) {
+  println BUILD_URL
+  def numberPR=sh (
+    script: "echo ${BRANCH_NAME} | cut -d '-' -f '2'",
+    returnStdout: true
+  ).trim()
+  def startIt='https://github.rackspace.com/api/v3/repos/<ORG_NAME_OR_USERNAME>/<REPO>/issues/'
```
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/Jenkinsfile]: I'm not sure what goes here. We may have to have the BPI team update this when they have their repo.

File: BPI-user-guide/doc/BPI-user-guide/internal/Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        returnStdout: true
+      ).trim()
+      println numberPR
+      isPR = env.BRANCH_NAME ? env.BRANCH_NAME.startsWith("PR-") : false
+      deleteDir() //clean out workspace for next run
+      echo "checking out"
+      // Checkout behavior explicit definition because the plugin downloads
+      // things differently depending on what it's looking at.
+      if (isPR) {
+        checkout scm:[$class: 'GitSCM', userRemoteConfigs: [[credentialsId: 'githubSvc', url: 'ssh://git@github.rackspace.com/<ORG_NAME_OR_USERNAME>/<REPO>.git']]]
```
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/Jenkinsfile]: I'm not sure what goes here. We may have to have the BPI team update this when they have their repo.

File: BPI-user-guide/doc/BPI-user-guide/internal/Jenkinsfile [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        sh "git fetch origin pull/'${numberPR}'/head:'${BRANCH_NAME}'"
+        sh "git checkout '${BRANCH_NAME}'"
+      } else {
+        checkout scm:[$class: 'GitSCM', userRemoteConfigs: [[credentialsId: 'githubSvc', url: 'git@github.rackspace.com:<ORG_NAME_OR_USERNAME>/<REPO>.git']]]
```
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/Jenkinsfile]: I'm not sure what goes here. We may have to have the BPI team update this when they have their repo.

File: BPI-user-guide/doc/BPI-user-guide/internal/Makefile0 (hunk 1 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/Makefile0]: We do need to have the capability to publish both internally and externally based on folder. The RPC repo does this: https://github.com/rackerlabs/docs-rpc/tree/master/doc

File: BPI-user-guide/doc/BPI-user-guide/internal/build.sh (hunk 2 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/build.sh]: I copied the whole file.

File: BPI-user-guide/doc/BPI-user-guide/internal/publish.sh (hunk 4 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/publish.sh]: I copied the whole file.

File: BPI-user-guide/doc/BPI-user-guide/internal/variables.sh (hunk 6 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/internal/variables.sh]: I copied the whole file.

File: BPI-user-guide/doc/BPI-user-guide/external/conf.py (hunk 7 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/external/conf.py]: Done.

File: BPI-user-guide/doc/BPI-user-guide/external/conf.py (hunk 8 above)
Comment (erikmwilson) [inline BPI-user-guide/doc/BPI-user-guide/external/conf.py]: Done.

**Comment by erikmwilson** [comment]
@nimbinatus Is the root directory supposed to be doc or docs? Most of our repos have doc as the folder name.

**Comment by klevenstein** [comment]
@erikmwilson possible dumb question, but is there a reason you're doing this here instead of in a new docs-bpi repo?

**Comment by nimbinatus** [comment]
@erikmwilson, for the directory name, it could be either doc or docs. The important point is that there is a separate directory in the root of the repo holding the content.

For the internal and external publishing question, no, the system is not currently set up to allow internal and external publishing from the same repo. If we wanted to do that, we need to host all content in GitHub Enterprise, and then we'll have to set up some additional logic in the Jenkinsfile to open up a PR on the external (rackerlabs) GitHub repo that will then need to be reviewed to ensure the documentation that gets published externally is truly external.

Cesar Garza asked for the ability to have one content repo publishing to two places like we used to have with Docbook, and I've added that requirement to the next iteration of the Jenkins system. I mentioned the idea that it would come into the external system as a PR and would therefore need a second review, and he seemed very happy with that.

**Comment by erikmwilson** [comment]
@klevenstein I figured we would get it to a working state and then move it over there. I guess we could have done it either way.

**Comment by erikmwilson** [comment]
@nimbinatus I'm confused about what we can do with our current system. In RPC, we currently publish both internal and external docs from the same repo (with some common content shared with both). We would like to do this with the BPI doc. Cesar also asked for conditional text (which would help with the above single sourcing) - but they are no the same thing.

**Comment by erikmwilson** [comment]
@nimbinatus Re: the doc/docs question. Do you have to change all the publishing files to make sure it is going to the correct folder (doc vs docs)? If so, the template publishing files should probably use doc since that seems to be what the majority of out repos use (at least the ones I looked at).

**Comment by nimbinatus** [comment]
@erikmwilson, the RPC system was put together as an interim solution until we had Jenkins internally. It is not a solution we can continue to use for at least a couple reasons. First, the solution was Sphinx specific, so it will not work with RAML or Markdown. Second, the solution relies on a Sphinx plugin that is developed by a volunteer rather than a Sphinx builtin function. There's no guarantee that solution will be viable long-term.

Going forward, we are planning to either have two separate repos or one repo on GitHub Enterprise that publishes both internally and externally. Cesar asked me both for conditional text and for single sourcing in two separate conversations, so both requests have been added as requirements to the Jenkins system going forward.

There are a couple places where we need to change the publishing files. The handbook repo used `docs` instead of `doc`, and that's what the RST version was set from. It might be worthwhile to put that directory name as a variable to be changed in the publishing files rather than hardcoding it all in.

**Comment by klevenstein** [comment]
@erikmwilson Trying to do non-template work in the template repo is only going to clutter up the template repo. I would suggest creating a new repo for the BPI content and closing this PR without merging it.

If I'm understanding @nimbinatus correctly, it sounds like what we ought to do is create the BPI repo over in [GitHub Enterprise](https://github.rackspace.com) instead of creating it here, if we're going to meet the internal/external publishing requirement.

**Comment by nimbinatus** [comment]
@klevenstein, yes, that's correct. My preference is housing the docs in the [IX](https://github.rackspace.com/IX/) org, but if BPI prefers to house the docs in their own repo, we can do that, too. Please let me know if they decide to host their own repo as we will need to hook them into Jenkins. Automatic linking to the Jenkins system from outside of IX is on the requirements for the next iteration of the internal publishing pipleine.

**Comment by erikmwilson** [comment]
@nimbinatus @klevenstein I think I was trying to shield the BPI team from seeing all the behind the scenes machinations as we got this together. I'm fine with working on it wherever it needs to be. Maybe we need a sandbox repo where we can work on things until they are ready to roll out (if we don't already have one).

**Comment by nimbinatus** [comment]
@erikmwilson, feel free to create a sandbox repo in IX on GitHub Enterprise! I removed all the create-repo restrictions that were there previously, so tinker away :)

**Comment by klevenstein** [comment]
@erikmwilson Yeah, I get what you're trying to do, and that makes sense. But we really shouldn't be doing anything in the template repo that isn't actually to do with fixing the templates. 

Let's set something up in https://github.rackspace.com/IX and work there. We can close this PR without merging it.

**Comment by erikmwilson** [comment]
I will move this over there. Do we need to have another meeting with Cesar (with IA and Tools) to let him know how things will be published (or does he already know)? I worry we may have overpromised what we can do based on the RPC model. And I still don't understand what we can't replicate that for the BPI docs.

**Comment by klevenstein** [comment]
@erikmwilson @nimbinatus Maybe it would be helpful for the three of us to have a quick meeting and make sure we're all on the same page. It looks like we've all been talking to Cesar separately, so we should sync up.

**Comment by klevenstein** [comment]
@erikmwilson whenever you've created the new BPI repo in GitHub Enterprise, go ahead and close this issue without merging.

**Comment by erikmwilson** [comment]
Moved to: https://github.rackspace.com/IX/docs-BPI/pull/1

**Pull request closed without merging**